assignments/week_4/week_4_exercises/exercise_4_1/exercise_4_1.md

```diff
@@ -0,0 +1,127 @@
+# Exercise 4.1
+This is the first exercise of week 4, and the first one involving ROS with C++!
+
+# Exercise Objective
+For this exercise you will be adding a subscriber to a ROS node. The end result will be a node which
+subscribes to `/string_topic`, and prints the messages data to the ROS console.
+
+Adding a subscriber involves:
+ 
+ * Including the correct message header
+ * Making a callback function
+ * Creating the subscriber object
+ 
+To test that your subscriber works, you will:
+ * Run the node using `roslaunch`
+ * Find the log file created when the node was running
+ * Turn in the node and log file to show your node worked
+ 
+# Message Headers
+To use message types, you need to include the correct header.
+In this exercise, we will use the `String` message type from the `std_msgs` package.
+To include a message, add an include message at the top of the file, like this:
+```c++
+#include <std_msgs/String.h>
+```
+
+# Making the callback function
+A callback function should have a `void` return type, and take in a single parameter
+with the type of message the subscriber is receiving. 
+```c++
+void callback_function(std_msgs::String string_msg)
+{
+    // LOGGING GOES HERE
+}
+```
+
+To print the string, you need to use `ROS_INFO_STREAM`, which takes a string input
+(or something which can easily be converted, like a number type) and prints it to the 
+ROS console. There's other variants like `ROS_DEBUG_STREAM` or `ROS_ERROR_STREAM`, but 
+for this exercise we just want to display a message, so we use `ROS_INFO_STREAM`.
+
+Also, be sure to remember that the String message object IS NOT a string object, to 
+access the actual string, you need to access it's `data` member variable.
+
+So the final callback function should look something like this:
+```c++
+void callback_function(std_msgs::String string_msg){
+    ROS_INFO_STREAM(string_msg.data);
+}
+```
+
+# Creating the subscriber object
+To create a subscriber, you will use the `subscribe` method of the `NodeHandle` object
+already created for you. The type of the object created will to be `ros::Subscriber`.
+```c++
+ros::Subscriber subscriber = p_nh.subscribe(...);
+```
+
+The arguments to this function are:
+* The topic we subscribe to, which for this exercise is `"/string_topic"`
+* The queue size we want, which should be 1
+* A reference to the callback function the subscriber should use. In C++, the way
+you reference a function is by using the `&` character right before the name of the 
+function, WITHOUT the parenthesis, otherwise the function will be called.
+
+In the end, your subscriber construction should look like this:
+```c++
+ros::Subscriber subscriber = p_nh.subscribe("/string_topic", 1, &callback_function);
+```
+
+# Compile the node and package
+To compile the `week_4_exercises` package, we need to create a workspace. 
```

Comment:
Should this just be cloned into the workspace they already have? I don't think we want multiple workspaces for them initially.